Cannot compile Clippy on rustc 1.23.0-nightly (d762b1d6c 2017-11-04)

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->

`clippy` does not compile on the current nightly, rustc 1.23.0-nightly (d762b1d6c 2017-11-04)

I see the following errors:

```

error[E0531]: cannot find tuple struct/variant `ItemDefaultImpl` in module `hir`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.168/src/utils/inspector.rs:406:14
    |
406 |         hir::ItemDefaultImpl(_, ref _trait_ref) => {
    |              ^^^^^^^^^^^^^^^ not found in `hir`

error[E0531]: cannot find tuple struct/variant `ItemDefaultImpl` in module `hir`
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.168/src/functions.rs:89:81
   |
89 |             matches!(item.node, hir::ItemImpl(_, _, _, _, Some(_), _, _) | hir::ItemDefaultImpl(..))
   |                                                                                 ^^^^^^^^^^^^^^^ not found in `hir`

error[E0531]: cannot find tuple struct/variant `ItemDefaultImpl` in module `hir`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.168/src/missing_doc.rs:135:18
    |
135 |             hir::ItemDefaultImpl(..) |
    |                  ^^^^^^^^^^^^^^^ not found in `hir`

error[E0599]: no method named `trait_has_default_impl` found for type `rustc::ty::TyCtxt<'_, '_, '_>` in the current scope
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.168/src/utils/inspector.rs:400:23
    |
400 |             if cx.tcx.trait_has_default_impl(did) {
    |                       ^^^^^^^^^^^^^^^^^^^^^^

error[E0023]: this pattern has 4 fields, but the corresponding tuple variant has 5 fields
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.168/src/len_zero.rs:71:13
   |
71 |             ItemTrait(_, _, _, ref trait_items) => check_trait_items(cx, item, trait_items),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 5 fields, found 4

error: aborting due to 5 previous errors

error: Could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `clippy v0.0.168`, intermediate artifacts can be found at `/tmp/cargo-install.tXHpQc2gbFKg`

Caused by:
  build failed

```

--- a/clippy_lints/src/functions.rs
+++ b/clippy_lints/src/functions.rs
@@ -86,7 +86,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
         use rustc::hir::map::Node::*;
 
         let is_impl = if let Some(NodeItem(item)) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(nodeid)) {
-            matches!(item.node, hir::ItemImpl(_, _, _, _, Some(_), _, _) | hir::ItemDefaultImpl(..))
+            matches!(item.node, hir::ItemImpl(_, _, _, _, Some(_), _, _) | hir::ItemAutoImpl(..))
         } else {
             false
         };

--- a/clippy_lints/src/len_zero.rs
+++ b/clippy_lints/src/len_zero.rs
@@ -68,7 +68,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LenZero {
         }
 
         match item.node {
-            ItemTrait(_, _, _, ref trait_items) => check_trait_items(cx, item, trait_items),
+            ItemTrait(_, _, _, _, ref trait_items) => check_trait_items(cx, item, trait_items),
             ItemImpl(_, _, _, _, None, _, ref impl_items) => check_impl_items(cx, item, impl_items),
             _ => (),
         }

--- a/clippy_lints/src/missing_doc.rs
+++ b/clippy_lints/src/missing_doc.rs
@@ -132,7 +132,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingDoc {
             hir::ItemGlobalAsm(..) => "an assembly blob",
             hir::ItemTy(..) => "a type alias",
             hir::ItemUnion(..) => "a union",
-            hir::ItemDefaultImpl(..) |
+            hir::ItemAutoImpl(..) |
             hir::ItemExternCrate(..) |
             hir::ItemForeignMod(..) |
             hir::ItemImpl(..) |

--- a/clippy_lints/src/needless_pass_by_value.rs
+++ b/clippy_lints/src/needless_pass_by_value.rs
@@ -87,7 +87,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NeedlessPassByValue {
 
         // Exclude non-inherent impls
         if let Some(NodeItem(item)) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(node_id)) {
-            if matches!(item.node, ItemImpl(_, _, _, _, Some(_), _, _) | ItemDefaultImpl(..)) {
+            if matches!(item.node, ItemImpl(_, _, _, _, Some(_), _, _) | ItemAutoImpl(..)) {
                 return;
             }
         }

--- a/clippy_lints/src/utils/inspector.rs
+++ b/clippy_lints/src/utils/inspector.rs
@@ -397,14 +397,14 @@ fn print_item(cx: &LateContext, item: &hir::Item) {
         },
         hir::ItemTrait(..) => {
             println!("trait decl");
-            if cx.tcx.trait_has_default_impl(did) {
-                println!("trait has a default impl");
+            if cx.tcx.trait_is_auto(did) {
+                println!("trait is auto");
             } else {
-                println!("trait has no default impl");
+                println!("trait is not auto");
             }
         },
-        hir::ItemDefaultImpl(_, ref _trait_ref) => {
-            println!("default impl");
+        hir::ItemAutoImpl(_, ref _trait_ref) => {
+            println!("auto impl");
         },
         hir::ItemImpl(_, _, _, _, Some(ref _trait_ref), _, _) => {
             println!("trait impl");

